### PR TITLE
Change implementation of extensions in ClientHello to a list - v2

### PR DESCRIPTION
**Pull request by tomato42**
As the name suggests, the extensions mechanism should be extensible this patch series makes it so.

Implements:
 - generic `TLSExtension` class (e.g. for use when the extension doesn't require a payload)
 - `SNIExtension` for handling server_name extension in full of its potential (support for more than one name and names different than domain names)
 - `NPNExtension` for handling the next protocol extension
 - `SRPExtension` for handling secure remote password
 - `ClientCertTypeExtension` for handling the cert_type extension (ClientHello variant)

Makes parsing and serializing extensions generic, moves it out of `ClientHello` class.

As a side effect, `ClientHello` now preserves the order of extensions, doesn't drop additional names from SNI and can also send and receive arbitrary extensions without any additional changes. Fully preserves old API using property mechanism.

Includes extension-specific test coverage for the introduced classes and some extra documentation for `ClientHello` class.

v2:
 - renamed `ClntCertTypeExtension` to `ClientCertTypeExtension`
 - removed `lambda`'s and `map`'s

obsoletes pull #33

### CODE REVIEW

**Comment by tomato42** [comment]
when can I expect the review to be ready?

I'd like to post similar changes to ServerHello, but they depend on this pull request

**Comment by trevp** [comment]
Pretty good, few nits:
 - probably rename get_extension to getExtension() or _getExtension(), for consistency with other method names
 - could you explain the "XXX backward compatibility" better?
 - I think tlsextension.py should probably be renamed to extensions.py, feels more consistent with alerts, messages, etc.
 - Could your description be a little clearer about what existing behavior is old that you preserved, and what is new?
 - the "properties" are a little weird as that's a language feature not previously used.  Did you consider just changing the method names to getX()/setX()?  Why did you decide to go with properties?

**Comment by tomato42** [comment]
* get_extension: good point, I'll change to getExtension() - it is useful for external callers so no reason to mark it with "_"
 * "XXX backwards compatibility" - the old code depends on fields having particular values even if they were not sent by server - cursory check indicated that it is used in other parts of code. I'd like to remove such logic from parser/serialiser classes, but that's for a different patch set.
 * extensions.py - agreed
 * basically all behaviour should be preserved, the only new thing for callers is the extensions field. The old stuff is everything that's exposed through properties
 * I went with properties so that the new code is a drop in replacement for old code. setters and getters are very unpythonic...

**Comment by tomato42** [comment]
one more thing regarding old vs new behaviour - all test cases should be passing after every individual commit and there are no changes to test cases in commit that changes ClientHello to use extensions (1fca0be)
